### PR TITLE
Update logic for comparing OpenMM XML patches for equivalence

### LINKED ISSUE #999: Non-identical patches are culled when writing OpenMM XML files
In https://github.com/pandegroup/openmm/issues/2139 we observed that non-identical patches were being culled when writing OpenMM XML files.

The [logic for comparing equivalent patches](https://github.com/ParmEd/ParmEd/blob/master/parmed/openmm/parameters.py#L205-L207) mistakenly only checked for equivalents of bonds to be added and atoms to be deleted, but missed checking the atoms to be added had equivalent charges. This caused some patches that were not identical (such as CHARMM36 patches `NTER` and `GLYP`) to mistakenly discard one of the patches.

I'll open a PR to fix this shortly.

### PR DESCRIPTION
**Pull request by jchodera**
Fixes #999 

This is a WIP. I'm trying to figure out why surprisingly few patches are being judged as equivalent now.

### CODE REVIEW

**Comment by jchodera** [comment]
Hm, I realize now there was *already* a scheme for eliminating duplicate patches using [this template hasher](https://github.com/ParmEd/ParmEd/blob/6d3fc9d1bb9af0383c28151de6f61da129b539e3/parmed/openmm/parameters.py#L486-L492) which was never extended to polyatomic patches.

**Comment by jchodera** [comment]
I think I've fixed this by fixing the template hasher. 

@swails: Any idea what a good set of tests might be? Would it be sufficient to process the CHARMM36 protein forcefield (which contains ACE, ACED, ACP, ACPD which are all considered duplicate patches to OpenMM) and ensure that only the first patch is retained when the `CharmmParameterSet` is converted to a `OpenMMParameterSet`, or is there something more elaborate you'd like to see here?

**Comment by swails** [comment]
That sounds reasonable.

File: parmed/openmm/parameters.py [hunk 1]
```diff
@@ -468,11 +469,26 @@ def _find_unused_types(self, skip_residues):
...
-            return hash((atom.atomic_number, atom.type, atom.charge))
-        # TODO implement hash for polyatomic residues
-        return id(residue)
+        """
+        Create a unique hash for each residue and patch template using only properties rendered to OpenMM ffxml.
+        """
+        hash_info = tuple()
+        # Sort tuples of atom properties by atom name
+        if len(residue.atoms) > 0:
+            hash_info += tuple(sorted( [(atom.name, atom.type, str(atom.charge)) for atom in residue.atoms], key=lambda tup: tup[0]))
```
Comment (swails) [inline parmed/openmm/parameters.py]: This is already the standard sort key, so the `key` keyword seems unnecessary to me.

**Comment by jchodera** [comment]
Thanks! I've fixed the redundant sort key issue and added that test!

**Comment by swails** [comment]
Did you know that `dict` now preserves order as of 3.6? https://news.ycombinator.com/item?id=12460936

**Comment by jchodera** [comment]
No! That's awesome.

**Comment by jchodera** [comment]
Any idea why Jenkins thinks the build is failing? I can't find any errors in the 3.5 log, and the output just ends in this mysterious incantation:
```
03:29:03 TOTAL                                                                                                                                                                     22794   2306    90%   
03:29:03 ++ coverage xml -o ../coverage.xml
03:29:04 No source for code: '/iscratch/jenkins/miniconda3/envs/jenkins-parmed-3.5/lib/python3.5/site-packages/ParmEd-3.0.3+38.ga946df58-py3.5-linux-x86_64.egg/parmed/__init__.py'.
03:29:04 Build step 'Execute shell' marked build as failure
03:29:04 [Cobertura] Publishing Cobertura coverage report...
03:29:04 
03:29:04 Finished: FAILURE
```

**Comment by swails** [comment]
I never got the Jenkins build working again. Someday I’ll make a Jenkinsfile to encode everything there. But until then, we just have to get Travis working.

**Pull request merged**